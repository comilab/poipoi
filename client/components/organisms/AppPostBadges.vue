<template lang="pug">
ul.flex
  li.mx-1(v-if="post.pinned")
    AppButton(
      type="badge",
      color="blue",
      size="sm",
      :icon="faThumbtack",
      pill,
      aria-label="固定"
    )
  li.mx-1(v-if="post.rating === 'nsfw'")
    AppButton(
      type="badge",
      color="red",
      size="sm",
      :icon="faExclamationTriangle",
      :breakpoint="breakpoint",
      pill
    )
      span(v-show="!iconOnly") 閲覧注意
  li.mx-1(v-else-if="post.rating === 'r18'")
    AppButton(
      type="badge",
      color="red",
      size="sm",
      icon="r18",
      :breakpoint="breakpoint",
      pill
    )
      span(v-show="!iconOnly") R-18
  li.mx-1(v-if="post.scope === 'password'")
    AppButton(
      type="badge",
      color="yellow",
      size="sm",
      :icon="faLock",
      :breakpoint="breakpoint",
      pill
    )
      span(v-show="!iconOnly") パスワード
  li.mx-1(v-else-if="post.scope === 'private'")
    AppButton(
      type="badge",
      color="gray",
      size="sm",
      :icon="faEyeSlash",
      :breakpoint="breakpoint",
      pill
    )
      span(v-show="!iconOnly") 非公開
  li.mx-1(v-if="isOutOfDate")
    AppButton(
      type="badge",
      color="gray",
      size="sm",
      :icon="faCalendarTimes",
      :breakpoint="breakpoint",
      pill
    )
      span(v-show="!iconOnly") 公開期間外
</template>

<script lang="ts">
import { defineComponent, computed } from 'nuxt-composition-api'
import dayjs from 'dayjs'
import { faThumbtack, faEyeSlash, faCalendarTimes, faExclamationTriangle, faLock } from '@fortawesome/free-solid-svg-icons'

import Post from '~/models/Post'
import AppButton from '~/components/atoms/AppButton.vue'

export default defineComponent({
  components: {
    AppButton
  },
  props: {
    post: {
      type: Post,
      required: true
    },
    iconOnly: {
      type: Boolean,
      required: false,
      default: false
    }
  },
  setup (props) {
    const isOutOfDate = computed(() => {
      return (
        (props.post.publishStart && dayjs().isBefore(props.post.publishStart)) ||
        (props.post.publishEnd && dayjs().isAfter(props.post.publishEnd))
      )
    })

    const breakpoint = computed(() => props.iconOnly ? null : 'md')

    return {
      isOutOfDate,
      breakpoint,
      faThumbtack,
      faEyeSlash,
      faCalendarTimes,
      faExclamationTriangle,
      faLock
    }
  }
})
</script>
