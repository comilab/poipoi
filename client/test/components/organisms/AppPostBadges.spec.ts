import { render } from '@testing-library/vue'
import dayjs from 'dayjs'
import faker from 'faker'

import AppPostBadges from '~/components/organisms/AppPostBadges.vue'
import postFactory from '~/test/factories/post'

describe('components/organisms/AppPostBadges', () => {
  let options: any

  beforeEach(() => {
    options = {
      props: {
        post: postFactory.build()
      },
      stubs: {
        FontAwesomeIcon: true
      }
    }
  })

  test('renders correctly', () => {
    options.props.post = postFactory.build({
      pinned: true,
      rating: 'nsfw',
      scope: 'public',
      publishStart: null,
      publishEnd: null
    })
    const { container } = render(AppPostBadges, options)

    expect(container).toMatchSnapshot()
  })

  test('pinned', async () => {
    options.props.post = postFactory.build({
      pinned: true
    })
    const { queryByLabelText, updateProps } = render(AppPostBadges, options)

    expect(queryByLabelText('固定')).toBeInTheDocument()

    await updateProps({
      post: postFactory.build({
        pinned: false
      })
    })
    expect(queryByLabelText('固定')).not.toBeInTheDocument()
  })

  describe('rating', () => {
    test('none', () => {
      options.props.post = postFactory.build({
        rating: null
      })
      const { queryByText } = render(AppPostBadges, options)

      expect(queryByText('閲覧注意')).not.toBeInTheDocument()
      expect(queryByText('R-18')).not.toBeInTheDocument()
    })

    test('nsfw', () => {
      options.props.post = postFactory.build({
        rating: 'nsfw'
      })
      const { queryByText } = render(AppPostBadges, options)

      expect(queryByText('閲覧注意')).toBeInTheDocument()
      expect(queryByText('R-18')).not.toBeInTheDocument()
    })

    test('r18', () => {
      options.props.post = postFactory.build({
        rating: 'r18'
      })
      const { queryByText } = render(AppPostBadges, options)

      expect(queryByText('閲覧注意')).not.toBeInTheDocument()
      expect(queryByText('R-18')).toBeInTheDocument()
    })
  })

  describe('scope', () => {
    test('public', () => {
      options.props.post = postFactory.build({
        scope: 'public'
      })
      const { queryByText } = render(AppPostBadges, options)

      expect(queryByText('パスワード')).not.toBeInTheDocument()
      expect(queryByText('非公開')).not.toBeInTheDocument()
    })

    test('password', () => {
      options.props.post = postFactory.build({
        scope: 'password'
      })
      const { queryByText } = render(AppPostBadges, options)

      expect(queryByText('パスワード')).toBeInTheDocument()
      expect(queryByText('非公開')).not.toBeInTheDocument()
    })

    test('private', () => {
      options.props.post = postFactory.build({
        scope: 'private'
      })
      const { queryByText } = render(AppPostBadges, options)

      expect(queryByText('パスワード')).not.toBeInTheDocument()
      expect(queryByText('非公開')).toBeInTheDocument()
    })
  })

  describe('公開期間外', () => {
    test('publishStartが現在以降なら表示', () => {
      options.props.post = postFactory.build({
        publishStart: dayjs(faker.date.future()),
        publishEnd: null
      })
      const { queryByText } = render(AppPostBadges, options)

      expect(queryByText('公開期間外')).toBeInTheDocument()
    })

    test('publishEndが現在以前なら表示', () => {
      options.props.post = postFactory.build({
        publishStart: null,
        publishEnd: dayjs(faker.date.past())
      })
      const { queryByText } = render(AppPostBadges, options)

      expect(queryByText('公開期間外')).toBeInTheDocument()
    })

    test.each([
      { publishStart: null, publishEnd: null },
      { publishStart: faker.date.past(), publishEnd: null },
      { publishStart: null, publishEnd: faker.date.future() }
    ])('それ以外なら表示しない', (attributes: any) => {
      options.props.post = postFactory.build(attributes)
      const { queryByText } = render(AppPostBadges, options)

      expect(queryByText('公開期間外')).not.toBeInTheDocument()
    })
  })
})
