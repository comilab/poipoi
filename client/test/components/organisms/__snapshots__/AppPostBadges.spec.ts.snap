// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/organisms/AppPostBadges renders correctly 1`] = `
<div>
  <ul
    class="flex"
  >
    <li
      class="mx-1"
    >
      <div
        aria-label="固定"
        class="inline-flex items-center justify-center whitespace-no-wrap transition-colors duration-300 disabled_opacity-75 disabled_pointer-events-none focus_outline-none pointer-events-none bg-blue-600 text-white hover_bg-blue-700 text-sm px-2 py-1 square-sm rounded-full"
        data-testid="button"
      >
        <div>
          <fontawesomeicon-stub
            class="fa-fw"
            data-testid="icon"
            icon="[object Object]"
          />
        </div>
        <div
          class=""
          data-testid="text"
        />
      </div>
    </li>
    <li
      class="mx-1"
    >
      <div
        class="inline-flex items-center justify-center whitespace-no-wrap transition-colors duration-300 disabled_opacity-75 disabled_pointer-events-none focus_outline-none pointer-events-none bg-red-700 text-white hover_bg-red-800 text-sm px-2 py-1 square-sm breakpoint-md rounded-full"
        data-testid="button"
      >
        <div>
          <fontawesomeicon-stub
            class="fa-fw md_mr-2"
            data-testid="icon"
            icon="[object Object]"
          />
        </div>
        <div
          class="hidden md_block"
          data-testid="text"
        >
          <span>
            閲覧注意
          </span>
        </div>
      </div>
    </li>
    <!---->
    <!---->
  </ul>
</div>
`;
